### PR TITLE
TFTRT: Add Sigmoid, Tanh ops and unit tests.

### PR DESCRIPTION
**Pull request by trevor-m**
Add conversion for all three activation types supported by TensorRT's IActivationlayer: Relu, Tanh, Sigmoid. Also adds unit tests for all three ops.

### CODE REVIEW

File: tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc [hunk 1]
```diff
@@ -1964,6 +1964,83 @@ TEST_F(OpConverterTest, ConvertRelu6) {
...
+  auto get_act_output = [](std::string op_name, float input) -> float {
+    if (op_name == "Relu") {
+      return (input > 0.0f) ? input : 0.0f;
+    } else if (op_name == "Sigmoid") {
+      return 1.0f / (1.0f + std::exp(-input));
+    } else if (op_name == "Tanh") {
+      return std::tanh(input);
+    }
+    EXPECT_TRUE(false);
+    return input;
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc]: nit: return 0.

File: tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc [hunk 2]
```diff
@@ -1964,6 +1964,83 @@ TEST_F(OpConverterTest, ConvertRelu6) {
...
+  };
+
+  {
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc]: nit: remove this outer curly braces.

File: tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc [hunk 3]
```diff
@@ -1964,6 +1964,83 @@ TEST_F(OpConverterTest, ConvertRelu6) {
...
+      EXPECT_TRUE(TrtDimsEqualsArray({1, 2, 3}, output.tensor()->getDimensions()))
+          << output.DebugString();
+
+      const std::vector<float> input_data = {-100, -2, -1, 0, 1, 100};
+      std::vector<float> output_data(6);
+      BuildAndRun("input", input_data, "my_act", &output_data);
+      for (int i = 0; i < input_data.size(); i++) {
+        const float expected_output = get_act_output(op_name, input_data[i]);
+        EXPECT_FLOAT_EQ(output_data[i], expected_output)
+          << op_name << "(" << input_data[i] << ") should be equal to "
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc]: We don't need these logs, if the check fails it should print similar logs.

File: tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc [hunk 4]
```diff
@@ -1964,6 +1964,83 @@ TEST_F(OpConverterTest, ConvertRelu6) {
...
+    // Ok.
+    for (std::string op_name : {"Relu", "Sigmoid", "Tanh"}) {
+      Reset();
+      NodeDef node_def = get_act_nodedef(op_name);
+      AddTestTensor("input", {1, 2, 3});
+      RunValidationAndConversion(node_def);
+      TRT_TensorOrWeights output;
+      TF_EXPECT_OK(GetTensorOrWeights("my_act", &output));
+      EXPECT_TRUE(output.is_tensor());
+      EXPECT_TRUE(TrtDimsEqualsArray({1, 2, 3}, output.tensor()->getDimensions()))
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/convert/convert_nodes_test.cc]: nit: use ExpectTrtDimsEqualsArray

**Review by aaroey — CHANGES_REQUESTED** [review]
Thanks for the PR!

**Comment by aaroey** [comment]
@trevor-m thanks for the fix. Would you please fix the clang-format problem as well?

**Comment by trevor-m** [comment]
Hey Lambda, thanks for the review.
Unfortunately we can't use ASSERT_TRUE(false) because the function has a non-void return value. See https://groups.google.com/a/chromium.org/forum/#!topic/chromium-dev/7clymoTb3A0
Let me know if you know a work around or better alternative than EXPECT_TRUE().

Also, is there a way I can run the clang-format check locally?

Thanks again!

**Comment by aaroey** [comment]
@trevor-m Thanks for the fixes. For formatting we can just run the clang-format tool, but that will change the untouched lines as well, and I don't think we have good solution to that.

**Comment by harshini-gadige** [comment]
> @trevor-m Thanks for the fixes. For formatting we can just run the clang-format tool, but that will change the untouched lines as well, and I don't think we have good solution to that.

@trevor-m  Any update please ?

**Comment by trevor-m** [comment]
> > @trevor-m Thanks for the fixes. For formatting we can just run the clang-format tool, but that will change the untouched lines as well, and I don't think we have good solution to that.
> 
> @trevor-m Any update please ?

Update on what?

**Comment by harshini-gadige** [comment]
> > > @trevor-m Thanks for the fixes. For formatting we can just run the clang-format tool, but that will change the untouched lines as well, and I don't think we have good solution to that.
> > 
> > 
> > @trevor-m Any update please ?
> 
> Update on what?

I see this PR has not been approved by @aaroey yet. So wanted to check if there are any changes pending from your side.

**Comment by trevor-m** [comment]
> > > > @trevor-m Thanks for the fixes. For formatting we can just run the clang-format tool, but that will change the untouched lines as well, and I don't think we have good solution to that.
> > > 
> > > 
> > > @trevor-m Any update please ?
> > 
> > 
> > Update on what?
> 
> I see this PR has not been approved by @aaroey yet. So wanted to check if there are any changes pending from your side.

Ah I see, I believe I have addressed all of @aaroey's requested changes.

**Comment by trevor-m** [comment]
> Hi @trevor-m, it looks like `binary_tensor_weight_broadcast_test` and `multi_connection_neighbor_engine_test` are failing, would you help to run them to make sure they are not broken?
> 
> Thanks.

Hey Lambda, it looks like those tests were using sigmoid to purposely break the graph into separate engines. I've updated them to use `sin` instead which we use in other tests for the same purpose.

File: tensorflow/contrib/tensorrt/test/binary_tensor_weight_broadcast_test.py [hunk 5]
```diff
@@ -56,10 +56,14 @@ def GetParams(self):
       ]:
         a = self._ConstOp(weights_shape)
         f = x + a
-        x = math_ops.sigmoid(f)
+        # sin is used to break the graph into separate segments since it is an
```
Comment (aaroey) [inline tensorflow/contrib/tensorrt/test/binary_tensor_weight_broadcast_test.py]: Please remove these comments and use `self.trt_incompatible_op()` instead.

**Review by aaroey — DISMISSED** [review]

**Comment by aaroey** [comment]
Thanks @trevor-m for the fix, I'll retry.

**Comment by aaroey** [comment]
Hi @trevor-m, sorry but we got one more errors:

```
tensorflow/contrib/tensorrt/test/binary_tensor_weight_broadcast_test.py:29: [W0611(unused-import), ] Unused math_ops imported from tensorflow.python.ops

tensorflow/contrib/tensorrt/test/multi_connection_neighbor_engine_test.py:28: [W0611(unused-import), ] Unused gen_math_ops imported from tensorflow.python.ops

tensorflow/contrib/tensorrt/test/multi_connection_neighbor_engine_test.py:29: [W0611(unused-import), ] Unused math_ops imported from tensorflow.python.ops
```

Would you help to fix them as well?
Thanks.

**Review by aaroey — APPROVED** [review]

**Pull request merged**